### ISSUE TITLE
rustc crashes with exit code 3221225501

### ISSUE THREAD

**Issue by torkleyy** [rust-lang/rust#42544]
Channels: `beta` and `nightly`

OS: Windows (AppVeyor CI)

```
Caused by:
  process didn't exit successfully: `rustc --crate-name dxguid C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\dxguid-sys-0.2.0\src\lib.rs --crate-type lib --emit=dep-info,link -C debuginfo=2 -C metadata=e6ed05839a70ac27 -C extra-filename=-e6ed05839a70ac27 --out-dir C:\projects\amethyst\target\debug\deps -L dependency=C:\projects\amethyst\target\debug\deps --extern winapi=C:\projects\amethyst\target\debug\deps\libwinapi-97963d09fc19ffa3.rlib --cap-lints allow` (exit code: 3221225501)
Build failed, waiting for other jobs to finish...
error: build failed
```

Complete build log: https://ci.appveyor.com/project/amethyst/amethyst/build/1.0.336/job/1gg0nmj81hj67t4s

**Comment by retep998** [comment]
Issue did not exist in `2017-05-21`. Issue *does* exist in `2017-06-07`.

Issue is that rustc is suddenly burning through a *ton* of memory. I had to kill it when it got to around 5GB because I don't want my computer to swap to death.

**Comment by retep998** [comment]
Based on the commits in the commit range, I'm going to accuse @jseyfried of causing this with https://github.com/rust-lang/rust/pull/40847

**Comment by retep998** [comment]
Some performance testing by removing lines indicates that the time taken and memory used is roughly linear with the number of lines. Approximately 0.1 seconds and 12MB of memory per line.

**Comment by retep998** [comment]
Interestingly, the issue only occurs when `DEFINE_GUID!` is imported via `#[macro_use(DEFINE_GUID)] extern crate winapi;`. If I define the macro locally, then the issue goes away.

**Comment by nikomatsakis** [comment]
triage: P-high

**Comment by nikomatsakis** [comment]
ping @pnkfelix -- this seems to be of pretty high importance. I'm not sure whether @jseyfried has time to make progress on it now, maybe you can look into it a bit?

**Comment by jseyfried** [comment]
Sorry for the delay getting to this -- I worked on this yesterday and couldn't repro. I tried on a Linux and a Windows system with a couple of recent nightlies. I have some extra time this week, so once I can repro I believe I can investigate and fix within a day.

**Comment by jseyfried** [comment]
Reproduced, working on a fix now.

**Comment by jseyfried** [comment]
Fixed in #42728.
